src/test-provider/output-analyzer.ts

```diff
@@ -0,0 +1,72 @@
+export interface TestRunActions {
+  failed?(testId: string, time: number, message: string): void;
+  passed?(testId: string, time: number): void;
+  skipped?(testId: string): void;
+}
+
+// e.g. Execution took (90ms|0.1s|etc.)
+const timeRegex = /Execution took (\d?\.?\d+)(ms|s)/;
```

Comment:
Did you consider getting this information via BSP instead of parsing it from stdout? The stdout format is unstable and could change easily.

Reply:
Are you referring to the [test notifications](https://build-server-protocol.github.io/docs/specification.html#test-notifications-1)? I thought I could use `buildTarget/test` to run tests but AFAIK it runs all tests from the given projects or at least the whole test suites, depending on implementation. Also `The server's implementation decides the granularity at which tests are reported` so it seemed to me like too many cases have to be handled.
If you're not referring to the test notifications please elaborate more, because maybe I missed something. Can I obtain this kind of information from BSP when I'm using DAP to run tests? Does BSP take part in communication after the DAP session is started? I thought that both answers are no, but if they're affirmative I will happily get rid off regexes :D

Reply:
The question is whether we can actually get that data from the test runners :thinking: I quickly checked and what we are currently doing is just sending back line by line via the output notification. @adpi2 do you know if it would be possible to have some kind of test harness to report success/failure/elapsed time ?

Reply:
This information is part of the sbt testing framework so it should be possible to forward it via BSP. We don’t need to use buildTarget/test since it doesn’t support debugging. Instead, the build server can just push notifications for the test events.

Reply:
Yes, I agree with @olafurpg, we don't need to use the `buildTarget/test` and we can push test notifications during the `debugSession\start` session but it is not yet implemented. In sbt test-agent we have something called [`ForkEnvent`](https://github.com/sbt/sbt/blob/1f4517d5cc5f6417c24fc3478819d41b548e1ba6/testing/agent/src/main/java/sbt/ForkMain.java#L446-L482) that could be forwarded as [BSP test notifications](https://build-server-protocol.github.io/docs/specification.html#test-notifications-1) to Metals. (It looks like the `TestStart` notification is missing in `ForkEvent` currently, but it could be added).

Reply:
BSP test notifications are a bit problematic, since they are not part of DAP and it would require Metals forward them as LSP notifications that would update a mutable state for the test session to use. Not sure if we would not get any issue from that approach, especially if anyone runs a couple of tests at the same time (we wouldn't be able to tell which session should be updated).

I wonder whether it would be ok to just to use output event with the `data` field containing json with the information from the TestEvent. This would all be done in the same session and would be simple to handle on all possible frontends.

https://microsoft.github.io/debug-adapter-protocol/specification#Events_Output